egs2/rats/spk1/local/rats_trial_prep.py

```diff
@@ -1,75 +1,104 @@
+    # Load all .flac files in the dev directory
+    l_dev = glob.glob(os.path.join(args.data_dir, "data/dev/audio/*/src/*.flac"))
 
+    # Build spk to utt dictionary more efficiently
     d_spk2utt = {}
     for line in l_dev:
-        f_name = line.strip().split("/")[-1]
-        dir = line.strip().split("/")[-2]
-
-        if dir == "src":
-            src, lng, ch_info = f_name.strip().split("_")
-        else:
-            src, trs, lng, ch_info = f_name.strip().split("_")
+        f_name = line.split("/")[-1]
+        src = f_name.split("_")[0]  # Extract src directly from file name
+        spk_id = d_src2spk.get(src, None)
 
-        spk_id = d_src2spk[src]
-        if spk_id not in d_spk2utt.keys():
-            d_spk2utt[spk_id] = []
+        if spk_id is not None:
+            d_spk2utt.setdefault(spk_id, []).append(line)
 
-        d_spk2utt[spk_id].append(line)
-
-    print("# of spk: {}".format(len(d_spk2utt.keys())))
-    print("# of utt: {}".format(len(l_dev)))
+    print(f"# of spk: {len(d_spk2utt)}")
+    print(f"# of utt: {len(l_dev)}")
 
     used_trials = set()
 
-    def is_unique_trial(class_idx, utt1, utt2):
-        trial = (class_idx, utt1, utt2)
-        reverse_trial = (class_idx, utt2, utt1)
-        if trial in used_trials or reverse_trial in used_trials:
-            return False
-        used_trials.add(trial)
-        return True
+    def is_unique_trial(trials):
+        """Check if the trial is unique and add to used trials."""
+        unique_trials = []
+        for trial in trials:
+            class_idx, utt1, utt2 = trial
+            reverse_trial = (class_idx, utt2, utt1)
+            if trial not in used_trials and reverse_trial not in used_trials:
+                used_trials.add(trial)
+                unique_trials.append(trial)
+        return unique_trials
+
+    spk_keys = list(d_spk2utt.keys())
+    dev_set = set(l_dev)
 
     with open(args.out, "w") as f_out:
-        for _ in range(len(d_src2spk.keys()) * 20):
-            spk = np.random.choice(list(d_spk2utt.keys()), size=1)[0]
-            class_idx = np.random.randint(2)
-
-            if class_idx == 0:
-                while True:
-                    utt1, utt2 = np.random.choice(
-                        list(d_spk2utt[spk]), size=2, replace=False
-                    )
-                    if is_unique_trial(class_idx, utt1, utt2):
-                        break
-            elif class_idx == 1:
-                while True:
-                    tmp_lines = list(set(l_dev) - set(d_spk2utt[spk]))
-                    utt1, utt2 = np.random.choice(tmp_lines, size=2, replace=False)
-                    if is_unique_trial(class_idx, utt1, utt2):
-                        break
-
-            f_out.write("{} {} {}\n".format(class_idx, utt1, utt2))
+        # Pre-generate the random speaker and class indices to reduce calls to random
+        n_trials = len(d_src2spk) * 20
```

Comment:
How about moving this to the beginning part so that others could potentially be aware and maybe play with this variable?